prow/interrupts/interrupts.go

```diff
@@ -0,0 +1,213 @@
+	logrus.WithField("signal", s).Info("Received signal.")
+	m.seenSignal = true
+	m.c.Broadcast()
+}
+
+// WaitForGracefulShutdown waits until all registered servers and workers
+// have had time to gracefully shut down, or times out. This function is
+// blocking.
+func (m *manager) WaitForGracefulShutdown() {
+	finished := make(chan struct{})
+	go func() {
+		defer close(finished)
+		if !m.seenSignal {
+			m.c.Wait()
+		}
+	}()
+	select {
+	case <-finished:
+		logrus.Info("All workers gracefully terminated, exiting.")
+	case <-time.After(1 * time.Minute):
+		logrus.Warn("Timed out waiting for workers to gracefully terminate, exiting.")
+	}
+}
+
+// Context returns a context that stays is cancelled when an interrupt hits.
+// Using this context is a weak guarantee that your work will finish before
+// process exit as callers cannot signal that they are finished. Prefer to use
+// Run().
+func (m *manager) Context() context.Context {
+	ctx, cancel := context.WithCancel(context.Background())
+	m.wg.Add(1)
+	go func() {
+		defer m.wg.Done()
+
+		m.c.L.Lock()
+		if !m.seenSignal {
+			m.c.Wait()
+		}
+		cancel()
+		m.c.L.Unlock()
+	}()
+
+	return ctx
+}
+
+// Run will do work until an interrupt is received, then signal the
+// worker. This function is not blocking. Callers are expected to exit
+// only after WaitForGracefulShutdown returns to ensure all workers have
+// had time to shut down. This is preferable to getting the raw Context
+// as we can ensure that the work is finished before releasing our share
+// of the wait group on shutdown.
+func (m *manager) Run(work func(ctx context.Context)) {
+	ctx, cancel := context.WithCancel(context.Background())
+	m.wg.Add(1)
+	go func() {
+		defer m.wg.Done()
+		work(ctx)
+	}()
+
+	m.c.L.Lock()
+	if !m.seenSignal {
+		m.c.Wait()
+	}
+	cancel()
+	m.c.L.Unlock()
+}
+
+// ListenAndServe runs the HTTP server and handles shutting it down
+// gracefully on interrupts. This function is not blocking. Callers
+// are expected to exit only after WaitForGracefulShutdown returns to
+// ensure all servers have had time to shut down.
+func (m *manager) ListenAndServe(server *http.Server, gracePeriod time.Duration) {
+	go m.shutdown(server, gracePeriod)
+	go func() {
+		logrus.WithError(server.ListenAndServe()).Info("Server exited.")
+	}()
+}
+
+// ListenAndServe runs the HTTP server and handles shutting it down
+// gracefully on interrupts. This function is not blocking. Callers
+// are expected to exit only after WaitForGracefulShutdown returns to
+// ensure all servers have had time to shut down.
+func (m *manager) ListenAndServeTLS(server *http.Server, certFile, keyFile string, gracePeriod time.Duration) {
+	go m.shutdown(server, gracePeriod)
+	go func() {
+		logrus.WithError(server.ListenAndServeTLS(certFile, keyFile)).Info("Server exited.")
+	}()
+}
+
+// shutdown will shut down the server when an interrupt is received
+func (m *manager) shutdown(server *http.Server, gracePeriod time.Duration) {
+	m.wg.Add(1)
```

Comment:
It isn't ever safe to `Add` to a waitgroup and call `Done` in the same goroutine. `Add` need to be called before the caller calls `go m.shutdown()`